File: Wire-iOS/Sources/Components/AccentColorChangeHandler.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "AccentColorChangeHandler.h"

#import "WireSyncEngine+iOS.h"
#import "UIColor+WAZExtensions.h"
#import "Wire-Swift.h"


@interface AccentColorChangeHandler () <ZMUserObserver>

@property (nonatomic, copy) AccentColorChangeHandlerBlock handlerBlock;
@property (atomic, unsafe_unretained) id observer;

@property (nonatomic, strong) ZMUser *selfUser;
@property (nonatomic) id userObserverToken;

- (id)initWithObserver:(id)observer handlerBlock:(AccentColorChangeHandlerBlock)changeHandler;

@end



@implementation AccentColorChangeHandler

+ (instancetype)addObserver:(id)observer handlerBlock:(AccentColorChangeHandlerBlock)changeHandler;
{
    return [[self alloc] initWithObserver:observer handlerBlock:changeHandler];
}

- (id)init;
{
    return nil;
}

- (id)initWithObserver:(id)observer handlerBlock:(AccentColorChangeHandlerBlock)changeHandler
{
    self = [super init];
    if (self != nil) {
        self.handlerBlock = changeHandler;
        self.observer = observer;
        
        self.selfUser = [ZMUser selfUser];
        if (nil != [ZMUserSession sharedSession]) {
            self.userObserverToken = [UserChangeInfo addObserver:self
                                                         forUser:self.selfUser
                                                     userSession:[ZMUserSession sharedSession]];
        }
    }
    return self;
}

- (void)dealloc
{
    self.observer = nil;
}

- (void)userDidChange:(UserChangeInfo *)change
{
    if (change.accentColorValueChanged) {
        self.handlerBlock([UIColor accentColor], self.observer);
    }
}

@end

File: wire-ios/Wire-iOS/Sources/Components/AccentColorChangeHandler.swift
//
// Wire
// Copyright (C) 2026 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import WireSyncEngine

final class AccentColorChangeHandler: UserObserving {

    typealias AccentColorChangeHandlerBlock = (_ newColor: ZMAccentColor?) -> Void
    private var handlerBlock: AccentColorChangeHandlerBlock?
    private var userObserverToken: NSObjectProtocol?

    static func addObserver(
        userSession: UserSession,
        handlerBlock changeHandler: @escaping AccentColorChangeHandlerBlock
    ) -> Self {
        self.init(handlerBlock: changeHandler, userSession: userSession)
    }

    init(
        handlerBlock changeHandler: @escaping AccentColorChangeHandlerBlock,
        userSession: UserSession
    ) {
        self.handlerBlock = changeHandler

        if let selfUser = SelfUser.provider?.providedSelfUser {
            self.userObserverToken = userSession.addUserObserver(self, for: selfUser)
        }
    }

    func userDidChange(_ change: UserChangeInfo) {
        if change.accentColorValueChanged {
            handlerBlock?(change.user.zmAccentColor)
        }
    }
}
